[inductor] in emulate_precision_casts, disable fma fusion in triton

v0i0 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #163073



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

v0i0 committed:
[inductor] in emulate_precision_casts, disable fma fusion in triton

[ghstack-poisoned]

v0i0 committed:
Update on "[inductor] in emulate_precision_casts, disable fma fusion in triton"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

eellison reviewed (commented):
Would you mind adding/extending existing tests to check that we do not add this flag to both:

- triton templates 
- user triton kernels

v0i0 committed:
Update

[ghstack-poisoned]

eellison reviewed (approved):
looks great, thanks!

v0i0 commented:
@StrongerXi I tracked the referenced issue by finding a roll-out where the delta is significant, then tracked the delta to the rope procedure, and added eager casts to match inductor behavior - that way i know this rounding is critical. Then i tried to get the same behavior from inductor emulate precision casts, and it didn't do it, even tho triton had the right casts in it. So i looked at the ptx, which contained fma. I knew that in nvcc `-fmad=false` disables implicit fma generation. Grepping through triton code then finds that flag in .../nvidia/.../compiler.py, mediated by `enable_fp_fusion`. So i added this change to thread it through, and the logprob error disappeared.

@eellison I added tests that mm triton templates & user triton code do not have that flag set, and a test that the fusion is correctly being avoided.

v0i0 commented:
@pytorchbot merge

eellison commented:
cc @v0i0 - this gives emulate_precision_casts its first real potential slowdown. is it worth it to config this  - potentially on by default ? thoughts ?

eellison reviewed (commented):
cc @jansel, @shunting314 for thoughts

jansel reviewed (approved):
Seems fine to me.

v0i0 committed:
Update on "[inductor] in emulate_precision_casts, disable fma fusion in triton"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

v0i0 committed:
Update

[ghstack-poisoned]
